pkg/volume/flexvolume/plugin-defaults.go

```diff
@@ -28,7 +32,20 @@ func logPrefix(plugin *flexVolumePlugin) string {
 	return "flexVolume driver " + plugin.driverName + ": "
 }
 
+func hash(path string) uint64 {
+	hash := fnv.New32a()
+	hashutil.DeepHashObject(hash, path)
+	return uint64(hash.Sum32())
+}
+
 func (plugin *pluginDefaults) GetVolumeName(spec *volume.Spec) (string, error) {
 	klog.V(4).Infof(logPrefix((*flexVolumePlugin)(plugin)), "using default GetVolumeName for volume ", spec.Name())
+	// GetVolumeName should return an unique volume name as possible
+	if spec.Volume != nil && spec.Volume.FlexVolume != nil {
+		options := spec.Volume.FlexVolume.Options
+		if options != nil && options["mountpath"] != "" {
```

Comment:
This expects that the flex driver uses `mountpath` option. This may work for your driver, but different flex drivers use different options. We should not hardcode any options here.

The best would be if [GetVolumeName](https://github.com/kubernetes/kubernetes/blob/11d83dfdc17f0a80a7da73accfd060084cfee0a1/pkg/volume/flexvolume/plugin.go#L125) used the value provided by the driver, but it was removed in https://github.com/kubernetes/kubernetes/pull/47400. See also https://github.com/kubernetes/kubernetes/issues/44737, suggested json metadata file was never implemented.

@chakri-nelluri, do you have any other suggestion?
cc @gnufied

Reply:
I passed `mountpath` as argument of `volume.Spec` in `plugin.ConstructVolumeSpec`
I think it works for all flex driver,  although it doesn't look elegant enough.

There is only one place to call `plugin. GetVolumeName` once  when kubelet starts
The volume name only works when the pod is `orphan pod`  which will be cleaned up in `reconciler.reconcile` loop.

kubelet/volumemanager/reconciler/reconciler.go  `syncStates`  =>  `reconstructVolume` = > volume/util/operationexecutor/operation_executor.go `ReconstructVolumeOperation` => `plugin.ConstructVolumeSpec`

Reply:
from code "unique volume name" is currently being generated anyway, and there is no semantic / convention that states `GetName()` should return something unique.

given how it's implemented now:
```
	var uniqueVolumeName v1.UniqueVolumeName
	if attachablePlugin != nil || deviceMountablePlugin != nil {
		uniqueVolumeName, err = util.GetUniqueVolumeNameFromSpec(plugin, volumeSpec)
		if err != nil {
			return nil, err
		}
	} else {
		uniqueVolumeName = util.GetUniqueVolumeNameFromSpecWithPod(volume.podName, plugin, volumeSpec)
	}
```

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L490

shall we just always use `GetUniqueVolumeNameFromSpecWithPod` to generate unique name for volume reconciliation? we can include pod namespace as well so it's guaranteed that volume name in the map is unique

Reply:
It will affect all plugins not only flex volume

Reply:
Options is not available in most of the code paths or when the mount is stale that is the reason GetVolumeName was disabled in the first place.
GetVolumeName should return unique name in the cluster except for inline volumes.
Are we seeing any issues after removing the prefix?